ui_framework/components/button/button.js

```diff
@@ -0,0 +1,135 @@
+import React, {
+  PropTypes,
+} from 'react';
+
+import classNames from 'classnames';
+import keyMirror from 'keymirror';
+
+import { KuiLoadingButtonIcon } from './button_icon';
+
+const KuiButton = props => {
+  if (props.isSubmit) {
+    // input is a void element tag and can't have children.
+    if ((props.children && typeof props.children !== 'string') || props.icon) {
+      throw new Error(
+        `KuiButton with isSubmit prop set to true can only accept string children, and can't
+        display icons. This is because input is a void element and can't contain children.`
+      );
+    }
+  }
+
+  function onClick() {
+    // Don't even trigger the onClick handler if we're disabled.
+    if (props.isDisabled) {
+      return;
+    }
+    props.onClick(props.data);
+  }
+
+  const icon =
+    props.isLoading
+    ? <KuiLoadingButtonIcon />
+    : props.icon;
+
+  const className = classNames('kuiButton', props.className, {
+    'kuiButton--iconText': icon,
+  });
+
+  let wrappedChildren;
+
+  if (props.children) {
+    // We need to wrap the text so that the icon's :first-child etc. seudo-selectors get applied
+    // correctly.
+    wrappedChildren = (
+      <span>{props.children}</span>
+    );
+  }
+
+  const elementType =
+    props.isSubmit
+    ? 'input'
+    : props.href
+    ? 'a'
+    : 'button';
+
+  const children =
+    props.isSubmit
+    ? null
+    : props.isIconOnRight
+    ? (
+      <span>
+        {wrappedChildren}
+        {icon}
+      </span>
+    ) : (
+      <span>
+        {icon}
+        {wrappedChildren}
+      </span>
+    );
+
+  return React.createElement(elementType, {
```

Comment:
Not a huge fan of `createElement`, _if_ it can be easily avoided. Usually I'd prefer an `if/else` that either renders an `input` or an `a`, as it usually makes the code easier to read and the component more easily understood from where it's used.

If I wrote this I'd probably create a `KibanaSubmitButton` and a `KibanaLinkButton` or something like that. Might be too many cases I'm not thinking about though. It just feels like there will be less edge-casing by going that route. (e.g. could we make more params `required` if we did that? or maybe each implementation becomes simpler? (so instead of if-ing based on params internally, you choose at the call-site (the location in code where a function is called) which "function" to call (aka component to use))